### PR TITLE
Conveyor/Conduit upgrade Assistance

### PR DESCRIPTION
**Pull request by ma44**
Crossed out no longer applies, assistance only works with rotatable blocks
~~Not really working well with walls due to some neat behaviors, see discord link for example
https://discordapp.com/channels/391020510269669376/396416151032299521/689548505311739977~~

This is a rehashing of #1009 but incorporates change requests by Anuken including a color change, dedicated keybinding and some code replacement

### CODE REVIEW

File: core/src/mindustry/input/Placement.java [hunk 1]
```diff
@@ -56,6 +56,30 @@
...
+    public static Array<Point2> followLineUntil(int startX, int startY, int endX, int endY){
+        Pools.freeAll(points);
+        points.clear();
+
+        Tile cursor = world.ltile(startX, startY);
+        Block filter = cursor.block();
+
+        while(true){
+            if(cursor.block() != filter) break;                        // keep looking until encountering a different block
+            if(points.contains(Pools.obtain(Point2.class, Point2::new).set(cursor.x, cursor.y))) break; // avoid biting its own tail
```
Comment (ma44) [inline core/src/mindustry/input/Placement.java]: While the original PR did have a change request to make this into a lambda statement, this utilizes a changing `cursor.x, cursor.y` making lambda expression usage error out, I don't have enough knowledge to figure out an alternative way of doing this so any suggestions would be helpful.

Comment (Anuken) [inline core/src/mindustry/input/Placement.java]:
```suggestion
            if(points.contains(p -> p.equals(cursor.x, cursor.y))) break; // avoid biting its own tail
```

File: core/src/mindustry/input/Placement.java [hunk 2]
```diff
@@ -56,6 +56,30 @@
...
+            points.add(Pools.obtain(Point2.class, Point2::new).set(cursor.x, cursor.y));
+            if(cursor.x == endX && cursor.y == endY) break;            // do not look further than the cursor
+            cursor = cursor.getNearby(cursor.rotation());              // move cursor to next block to check
+        }
+        // if the end is nowhere on the line, don't return anything
+
+        if(!points.contains(p -> p.equals((Pools.obtain(Point2.class, Point2::new).set(endX, endY))))){
```
Comment (Anuken) [inline core/src/mindustry/input/Placement.java]:
```suggestion
        if(!points.contains(p -> p.equals(endX, endY))){
```

File: core/src/mindustry/input/Placement.java [hunk 3]
```diff
@@ -56,6 +56,30 @@
...
+
+        Tile cursor = world.ltile(startX, startY);
+        Block filter = cursor.block();
+
+        while(true){
+            if(cursor.block() != filter) break;                        // keep looking until encountering a different block
+            if(points.contains(Pools.obtain(Point2.class, Point2::new).set(cursor.x, cursor.y))) break; // avoid biting its own tail
+            points.add(Pools.obtain(Point2.class, Point2::new).set(cursor.x, cursor.y));
+            if(cursor.x == endX && cursor.y == endY) break;            // do not look further than the cursor
+            cursor = cursor.getNearby(cursor.rotation());              // move cursor to next block to check
```
Comment (Anuken) [inline core/src/mindustry/input/Placement.java]: I don't think blocks without rotation should be part of the line. Even if this works with walls, it's confusing and unintuitive.

File: core/src/mindustry/input/Placement.java [hunk 4]
```diff
@@ -56,6 +56,30 @@
...
+        }
+        // if the end is nowhere on the line, don't return anything
+
+        if(!points.contains(p -> p.equals((Pools.obtain(Point2.class, Point2::new).set(endX, endY))))){
+            Pools.freeAll(points);
+            points.clear();
+            points.add(new Point2(startX, startY));
```
Comment (Anuken) [inline core/src/mindustry/input/Placement.java]:
```suggestion
            points.add(Pools.obtain(Point2.class, Point2::new).set(startX, startY));
```

File: core/src/mindustry/input/Placement.java (hunk 1 above)
Comment (ma44) [reply core/src/mindustry/input/Placement.java]: bloop
Fails to compile

File: core/src/mindustry/input/Placement.java (hunk 3 above)
Comment (ma44) [reply core/src/mindustry/input/Placement.java]: Agree with this, added a check to see if the block can rotate or not so only conveyors/conduits get changed

File: core/src/mindustry/input/Placement.java (hunk 1 above)
Comment (ma44) [reply core/src/mindustry/input/Placement.java]: Addressed

**Comment by Anuken** [comment]
I won't be merging this after all - the keybinds are already very bloated, and it's usually not too hard to just upgrade conveyors manually. I may revisit this if someone comes up with a way to do it intuitively without new keybinds.

**Pull request closed without merging**